### PR TITLE
fix(celery): close `celery.apply` spans even without after_task_publish, when using apply_async

### PR DESCRIPTION
**Pull request by wantsui**
The instrumentation for the Celery integration relies on various [Celery signals ](https://docs.celeryq.dev/en/stable/userguide/signals.html) in order to start and end the span when calling on `apply_async`.

The integration can fail if the expected signals don't trigger, which can lead to broken context propagation (and unexpected traces).

**Example:**
- dd-trace-py expects the signal `before_task_publish` to start the span then `after_task_publish` to close the span. If the `after_task_publish` signal never gets called (which can happen if a Celery exception occurs while processing the app), then the span won't finish.
- The same thing above can also happen to `task_prerun` and `task_postrun`.

**Solution**

This PR patches `apply_async` so that there is a check to see if there is a span lingering around and closes it when `apply_task` is called.

If an internal exception happens, the error will be marked on the `celery.apply` span.

To track this, I added new logs in debug mode:
> The after_task_publish signal was not called, so manually closing span

and 
> The task_postrun signal was not called, so manually closing span

There's a related PR https://github.com/DataDog/dd-trace-py/pull/10848 that works to improve how we extract information based on the protocols, that also affects when spans get closed or not.

Special Thanks:
- Thanks to @tabgok for going through this with me in great detail!
- @timmc-edx for helping us track it down!

[APMS-13158]

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

APMS-13158

[APMS-13158]: https://datadoghq.atlassian.net/browse/APMS-13158?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ddtrace/contrib/internal/celery/app.py [hunk 1]
```diff
@@ -96,3 +105,27 @@ def _traced_beat_inner(func, instance, args, kwargs):
             return func(*args, **kwargs)
 
     return _traced_beat_inner
+
+
+def _traced_apply_async_function(integration_config, fn_name, resource_fn=None):
```
Comment (tabgok) [inline ddtrace/contrib/internal/celery/app.py:110]: It likely won't be apparent to future readers why we're grabbing a span out of seemingly thin air and then closing it here. 

This is a perfect time for a comment explaining WHY we're doing this.  I see a good blurb in the test, but it should be added here as well.  I'd include something along the lines of "apply_async synchronously calls the before/after signals, so we make sure we close at the end"

File: releasenotes/notes/fix-celery-apply-async-span-close-b7a8db188459f5b5.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    tracing(celery): Fixes an issue where `celery.apply` spans didn't close if the after_task_publish or task_postrun signals didn't get called.
```
Comment (tabgok) [inline releasenotes/notes/fix-celery-apply-async-span-close-b7a8db188459f5b5.yaml:4]: nit: would add a little bit about why they weren't called (exception generated in handling)

**Review by tabgok — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/contrib/internal/celery/utils.py [hunk 3]
```diff
@@ -128,9 +128,12 @@ def retrieve_task_id(context):
...
     body = context.get("body")
-    if headers:
+    # Check if the id is in the headers, then check the body for it.
+    # If we don't check the id first, we could wrongly assume no task_id
+    # when the task_id is in the body.
+    if headers and 'id' in headers:
         # Protocol Version 2 (default from Celery 4.0)
         return headers.get("id")
-    else:
+    elif body and 'id' in body:
```
Comment (wantsui) [inline ddtrace/contrib/internal/celery/utils.py:137]: @tabgok I believe this is the place we failed to extract ids correctly with the old logic, so it's another reason why the span doesn't get closed.

Comment (wantsui) [reply ddtrace/contrib/internal/celery/utils.py]: Update: I'll resolve this issue in a separate PR: https://github.com/DataDog/dd-trace-py/pull/10848

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/contrib/internal/celery/utils.py [hunk 4]
```diff
@@ -131,6 +131,6 @@ def retrieve_task_id(context):
     if headers:
         # Protocol Version 2 (default from Celery 4.0)
         return headers.get("id")
-    else:
+    elif body:
```
[the remark this answers is not in the document]
Comment (wantsui) [reply ddtrace/contrib/internal/celery/utils.py]: That makes sense, https://github.com/DataDog/dd-trace-py/pull/10848 ended up getting this logic (and another fix) so I removed the change in utils from this PR.

File: releasenotes/notes/fix-celery-apply-async-span-close-b7a8db188459f5b5.yaml (hunk 2 above)
Comment (wantsui) [reply releasenotes/notes/fix-celery-apply-async-span-close-b7a8db188459f5b5.yaml]: fixed in the latest version of the release notes

**Review by tabgok — APPROVED** [review]

**Comment by iherasymenko** [comment]
Linking to https://github.com/DataDog/dd-trace-py/issues/11624 as this PR introduced a regression.

**Pull request merged**